pkg/manager/member/pvc_resizer.go

```diff
@@ -80,215 +79,361 @@ type pvcResizer struct {
 		if quantity, err := resource.ParseQuantity(dc.Spec.Worker.StorageSize); err == nil {
-			dmWorkerMemberType := v1alpha1.DMWorkerMemberType.String()
-			key := fmt.Sprintf("%s-%s-%s", dmWorkerMemberType, dc.Name, dmWorkerMemberType)
-			pvcPrefix2Quantity[key] = quantity
-		}
-		if err := p.patchPVCs(ns, selector.Add(*dmWorkerRequirement), pvcPrefix2Quantity); err != nil {
-			return err
+			ctx.desiredVolumeQuantity[v1alpha1.GetStorageVolumeName("", v1alpha1.DMWorkerMemberType)] = quantity
 		}
+	default:
+		return nil, fmt.Errorf("unsupported member type %s", comp)
 	}
-	return nil
-}
 
-func (p *pvcResizer) isVolumeExpansionSupported(storageClassName string) (bool, error) {
-	sc, err := p.deps.StorageClassLister.Get(storageClassName)
+	podVolumes, err := p.collectAcutalStatus(ns, ctx.selector)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
-	if sc.AllowVolumeExpansion == nil {
-		return false, nil
-	}
-	return *sc.AllowVolumeExpansion, nil
+	ctx.actualPodVolumes = podVolumes
+
+	return ctx, nil
 }
 
-// patchPVCs patches PVCs filtered by selector and prefix.
-func (p *pvcResizer) patchPVCs(ns string, selector labels.Selector, pvcQuantityInSpec map[string]resource.Quantity) error {
-	if len(pvcQuantityInSpec) == 0 {
-		return nil
+// updateVolumeStatus build volume status from `actualPodVolumes` and call the `updateVolumeStatusFn` function.
+func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
+	if ctx.sourceVolumeStatus == nil {
+		return
 	}
-	pvcs, err := p.deps.PVCLister.PersistentVolumeClaims(ns).List(selector)
-	if err != nil {
-		return err
+
+	getCapacity := func(volName v1alpha1.StorageVolumeName, pvc *corev1.PersistentVolumeClaim) (resource.Quantity, resource.Quantity, bool) {
+		var desired, actual resource.Quantity
+		var exist bool
+
+		if pvc.Status.Phase != corev1.ClaimBound {
+			return desired, actual, false
+		}
+		desired, exist = ctx.desiredVolumeQuantity[volName]
+		if !exist {
+			return desired, actual, false
+		}
+		actual, exist = pvc.Status.Capacity[corev1.ResourceStorage]
+		if !exist {
+			return desired, actual, false
+		}
+
+		return desired, actual, true
+	}
+
+	allStatus := map[v1alpha1.StorageVolumeName]*v1alpha1.StorageVolumeStatus{}
+	for _, podVolume := range ctx.actualPodVolumes {
+		for volName, pvc := range podVolume.volToPVCs {
+			desiredQuantity, actualQuantity, pred := getCapacity(volName, pvc)
+			if !pred {
+				continue
+			}
+
+			status, exist := allStatus[volName]
+			if !exist {
+				status = &v1alpha1.StorageVolumeStatus{
+					Name: volName,
+					ObservedStorageVolumeStatus: v1alpha1.ObservedStorageVolumeStatus{
+						BoundCount:      0,
+						CurrentCount:    0,
+						ResizedCount:    0,
+						CurrentCapacity: desiredQuantity,
+						ResizedCapacity: desiredQuantity,
+					},
+				}
+				allStatus[volName] = status
+			}
+
+			status.BoundCount++
+			if actualQuantity.Cmp(desiredQuantity) == 0 {
+				status.ResizedCount++
```

Comment:
Should the currentCount be increased too?

Reply:
No. Now `resizedCount` is the count for `desiredQuantity` and `currentCount` is the count for other.
If all volumes are resized, `currentCount` will be set to `resizedCount` in L364.